test/extensions/filters/http/composite/composite_filter_integration_test.cc

```diff
@@ -66,15 +147,77 @@ TEST_P(CompositeFilterIntegrationTest, TestBasic) {
   }
 
   {
-    const Http::TestRequestHeaderMapImpl request_headers = {{":method", "GET"},
-                                                            {":path", "/somepath"},
-                                                            {":scheme", "http"},
-                                                            {"match-header", "match"},
-                                                            {":authority", "blah"}};
-    auto response = codec_client_->makeRequestWithBody(request_headers, 1024);
+    auto response = codec_client_->makeRequestWithBody(match_request_headers_, 1024);
     ASSERT_TRUE(response->waitForEndStream());
     EXPECT_THAT(response->headers(), Http::HttpStatusIs("403"));
   }
 }
 
+// Verifies function of the per-route config in the ExtensionWithMatcher class.
+TEST_P(CompositeFilterIntegrationTest, TestPerRoute) {
+  prependCompositeFilter();
+  addPerRouteResponseCodeFilter(/*filter_name=*/"composite", /*route_prefix=*/"/somepath",
+                                /*code=*/401);
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  auto response = codec_client_->makeRequestWithBody(match_request_headers_, 1024);
+  ASSERT_TRUE(response->waitForEndStream());
+  EXPECT_THAT(response->headers(), Http::HttpStatusIs("401"));
+}
+
+// Test an empty match tree resolving with a per route config.
+TEST_P(CompositeFilterIntegrationTest, TestPerRouteEmptyMatcher) {
+  config_helper_.prependFilter(R"EOF(
+      name: composite
+      typed_config:
+        "@type": type.googleapis.com/envoy.extensions.common.matching.v3.ExtensionWithMatcher
+        extension_config:
+          name: composite
+          typed_config:
+            "@type": type.googleapis.com/envoy.extensions.filters.http.composite.v3.Composite
+    )EOF");
+  addPerRouteResponseCodeFilter(/*filter_name=*/"composite", /*route_prefix=*/"/somepath",
+                                /*code=*/402);
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  auto response = codec_client_->makeRequestWithBody(match_request_headers_, 1024);
+  ASSERT_TRUE(response->waitForEndStream());
+  EXPECT_THAT(response->headers(), Http::HttpStatusIs("402"));
+}
+
+// Test that the specified filters apply per route configs to requests.
+TEST_P(CompositeFilterIntegrationTest, TestPerRouteMultipleFilters) {
+  prependCompositeFilter(/*name=*/"composite_2");
```

Comment:
It is ok to use this function for **no per route** test(`TestBasic`)
But for **per route tests , this one especially**.  we don't want to use this function. We want to use filter config that doesn't have matcher. 

So config should looks like below (similar to what you have in `TestPerRouteEmptyMatcher`) 

```
http_filters {
      name: "composite"
      typed_config {
        [type.googleapis.com/envoy.extensions.common.matching.v3.ExtensionWithMatcher] {
          extension_config {
            name: "filter_name_1"
            typed_config {
              [type.googleapis.com/envoy.extensions.filters.http.composite.v3.Composite] {
              }
            }
          }
        }
      }
    }
    http_filters {
      name: "filter_name_2"
      typed_config {
        [type.googleapis.com/envoy.extensions.common.matching.v3.ExtensionWithMatcher] {
          extension_config {
            name: "composite_2"
            typed_config {
              [type.googleapis.com/envoy.extensions.filters.http.composite.v3.Composite] {
              }
            }
          }
        }
      }
    }
```